fvm/interface/BaseInterface.py

```diff
@@ -0,0 +1,135 @@
+import sys
+import numpy
+
+from fvm.utils import norm
+
+from fvm import Discretization, CylindricalDiscretization
+
+
+class BaseInterface:
+    '''This class defines a base interface to the NumPy backend for the
+    discretization. We use this so we can write higher level methods
+    such as pseudo-arclength continuation without knowing anything
+    about the underlying methods such as the solvers that are present
+    in the backend we are interfacing with.'''
+
+    def __init__(self, parameters, nx, ny, nz, dim, dof, x=None, y=None, z=None):
+        self.nx = nx
+        self.ny = ny
+        self.nz = nz
+        self.dim = dim
+        self.dof = dof
+
+        self.discretization = Discretization(parameters, nx, ny, nz, dim, dof, x, y, z)
+
+        if 'Problem Type' in parameters and 'Taylor-Couette' in parameters['Problem Type']:
+            self.discretization = CylindricalDiscretization(parameters, nx, ny, nz, dim, dof, x, y, z)
+
+        self.parameters = parameters
+
+    def debug_print(self, *args):
+        if self.parameters.get('Verbose', False):
+            print('Debug:', *args)
+            sys.stdout.flush()
+
+    def debug_print_residual(self, string, jac, x, rhs):
+        if self.parameters.get('Verbose', False):
+            r = norm(jac @ x - rhs)
+            self.debug_print(string, '{}'.format(r))
+
+    def set_parameter(self, name, value):
```

Comment:
This was on purpose so people don't change the value in the parameters dict that gets passed to the constructor. If they do that that won't work. Since there are many parameters, I want them to be able to set them through there at least initially.

Allowing them to set parameters through a setter makes it even more confusing in my opinion. If there is a way to watch for changes in the parameter dict I could use that instead.